Op - add interface for multigrid level creation

jeremylt opened the pull request:
This is a WIP to simplify the creation of multigrid operators (prolong, restrict, coarse) from a fine grid operator. Implementation should be straightforward. Refactoring the PETSc code might be a (mild) pain in the neck.

jeremylt commented:
I need to add tests and add this functionality to our multigrid and solids examples, but this should be pretty close to what I'm picturing for the interface.

I am expecting the user to provide
1) Coarse grid restriction (100% not libCEED's business)
2) Coarse to fine grid interpolation matrix (or enough info for me to make it)
3) Coarse grid basis (or enough info for me to make it)

From there I provide
1) Coarse grid operator (same quadrature space as fine grid)
2) Prolongation operator
3) Restriction operator

jeremylt commented:
The open ToDo is adding tests for the other 2 constructors and adding this to the Python interface.

jeremylt commented:
Ok, all new code is tested. The two remaining pieces are
- Decide the best way to scale for multiplicity
- Add functions to python interface

jeremylt commented:
ToDo Update:
- [ ] Clean up unused code in solids example
- [ ] Add functions to Python interface

@jedbrown, in my mind the new function interface is stable on the C/Fortran side. I'm not going to tag 'In-Review' until I finish the above ToDo, but if you have any comments on the interface, please let me know if you get a moment.

valeriabarra commented on gallery/multigrid/ceed-multigrid1.h:
```diff
@@ -0,0 +1,43 @@
+CEED_QFUNCTION(Multigrid1)(void *ctx, const CeedInt Q,
+                           const CeedScalar *const *in,
+                           CeedScalar *const *out) {
+  // in[0] is input, size (Q*1)
+  // in[1] is multiplicity, size (Q)
```
slightly inconsistent comment

jeremylt replied:
Updated

valeriabarra commented on tests/python/test-5-operator.py:
```diff
@@ -1551,6 +1551,342 @@ def test_533(ceed_resource):
+    # Create multigrid level
+    p_mult_fine = ceed.Vector(ncomp * nu_fine)
+    p_mult_fine.set_value(1.0)
+    [op_mass_coarse, op_prolong, op_restrict] = op_mass_fine.multigrid_create_tensor_h1_lagrange(p_mult_fine,
+                                                                                                 ru_coarse, p_coarse)
+
+    # Apply coarse mass matrix
+    u_coarse.set_value(1.0)
+    op_mass_coarse.apply(u_coarse, v_coarse)
+
+    # Check
+    with v_coarse.array_read() as v_array:
+        total = 0.0
+        for i in range(nu_coarse * ncomp):
+            total = total + v_array[i]
+        assert abs(total - 2.0) < 1E-13
+
+    # Prolong coarse u
+    op_prolong.apply(u_coarse, u_fine)
+
+    # Apply mass matrix
+    op_mass_fine.apply(u_fine, v_fine)
+
+    # Check
+    with v_fine.array_read() as v_array:
+        total = 0.0
+        for i in range(nu_fine * ncomp):
+            total = total + v_array[i]
+        assert abs(total - 2.0) < 1E-13
+
+    # Restrict state to coarse grid
+    op_restrict.apply(v_fine, v_coarse)
+
+    # Check
+    with v_coarse.array_read() as v_array:
+        total = 0.0
+        for i in range(nu_coarse * ncomp):
+            total = total + v_array[i]
+        assert abs(total - 2.0) < 1E-13
+
```
I'm a bit puzzled... what is the difference between this and the previous test? They have the same name and description

jeremylt replied:
Weird, removed

valeriabarra replied:
No problem! I just wanted to make sure I was not seeing double on my screen! 😅😂

valeriabarra commented:
LGTM overall. One question: I guess the case `size = 2` was not relevant for the gallery?

jeremylt commented:
I didn't add size=2 because we don't have any current two field multigrid examples. I was thinking that we can easily add as needed but I didn't see a good reason to add it before we needed it

valeriabarra commented:
I think for completeness sake it would be better and more consistent to have it, since we have already in the gallery `mass1d`, `mass2d`, `mass3d`, and so on for the Poisson operator.

jeremylt commented:
Sure. I added the case and (hopefully) expanded the test coverage to handle this new case.

jedbrown commented on interface/ceed-qfunction.c:
```diff
@@ -743,7 +743,7 @@ int CeedQFunctionDestroy(CeedQFunction *qf) {
   ierr = CeedFree(&(*qf)->outputfields); CeedChk(ierr);
   // Free ctx if identity
-  if ((*qf)->identity) {
+  if ((*qf)->ownctx) {
```
Should we use a similar pattern to `CeedVector`, where we hold the ownership pointer separately, instead of a `bool`?

jeremylt commented on interface/ceed-operator.c:
```diff
@@ -983,6 +1194,183 @@ int CeedOperatorLinearAssembleAddPointBlockDiagonal(CeedOperator op,
+**/
+int CeedOperatorMultigridLevelCreateTensorH1(CeedVector PMultFine,
+    CeedElemRestriction rstrCoarse, CeedBasis basisCoarse,
+    const CeedScalar *interpCtoF, CeedOperator opFine,
```
No reason. I agree its more natural to have opFine first.

jeremylt commented on interface/ceed-operator.c:
```diff
@@ -197,6 +197,211 @@ int CeedOperatorSingleView(CeedOperator op, bool sub, FILE *stream) {
+
+  @ ref Developer
+**/
+static inline int CeedOperatorGetActiveBasis(CeedOperator op,
```
Good point. Updated.

jedbrown commented on gallery/multigrid/ceed-multigrid2.h:
```diff
@@ -0,0 +1,47 @@
+  // in[1] is multiplicity, size (Q*2)
+  const CeedScalar *input = in[0];
+  const CeedScalar *mult = in[1];
+  // out[0] is output, size (Q*1)
```
```suggestion
  // out[0] is output, size (Q*2)
```

jeremylt commented:
I can't force an operator to call the QFunction with Q*size instead of Q (in our current interface), and copying the context slows down the cuda backends, which is why I made the 123 versions.

I agree about the naming.

I considered adding a CeedVectorReciprocal to make the QFunction simpler. That would be easy to add.

jedbrown commented:
Hmm, could probably abuse the `void *` to pass the integer, but I guess that's okay. It's a one-time cost, right? I'd do the reciprocal and then just call these `gallery/scale/` or something like that (because it's all they do).

jeremylt commented:
Because context can change between operator applications, we have to pay the cost to transfer `scale` on every application

jeremylt commented on interface/ceed-operator.c:
```diff
@@ -983,6 +1194,183 @@ int CeedOperatorLinearAssembleAddPointBlockDiagonal(CeedOperator op,
+
+  @param[in] PMultFine    L-vector multiplicity in parallel gather/scatter
+  @param[in] rstrCoarse   Coarse grid restriction
+  @param[in] degreeCoarse Coarse grid basis polynomial order
```
I am happy with the other changes - this should be the last open question, I believe.

For the MFEM integration, I'll want to use their 'transfer matrix', so I think that `CeedOperatorMultigridLevelCreateTensorH1` and `CeedOperatorMultigridLevelCreateH1` are what I will use for that.

That leaves replacing `CeedOperatorMultigridLevelCreateTensorH1Lagrange` with `CeedOperatorMultigridLevelCreate` where the transfer basis is automatically created out of the fine and coarse bases. It would be pretty flexible and would cover non-tensor and tensor bases. I think I should be able to use some of the same code in GetCollocatedGrad.

jeremylt replied:
Update - should be ready to go now

jeremylt commented:
Note - Travis currently failing due to some problem I can't reproduce locally in t202 with destroying the vectors. Will investigate more tomorrow.

jedbrown commented:
Surely we could track whether the context has actually changed. Maybe hash it on the CPU so we have something easy to compare with, or require that the user notify after changes.

jeremylt commented:
We have an issue discussing this, which I was alluding to above.

In theory the user or the QFunction can update the context, so context requires two copies, host->device and device->host. (not sure if the second one is actually in the GPU backends) 

In practice our context has been largely static, so we discussed having two types of context, static and dynamic. I could remove the 123 variants and followup with a separate PR to finally update our context like we discussed in Dec 2018.

jeremylt commented:
Note: need to fix reciprocal, none sync makes no sense

jedbrown commented:
I'm not sure I follow about the sync. I'd expect it should not need to eagerly sync because the next usage is usually still on the device.

jeremylt commented:
That was a note to future me to fix the cuda/hip backend reciprocal function because taking the reciprocal of a vector that has a vector in state CEED_CUDA/HIP_SYNC_NONE will result in gibberish.

jeremylt commented:
Ok, all fixed and ready to merge.

I added the new interface that creates the prolongation basis from the fine and coarse grid interpolation, and I fixed the couple of small issues I made a note about above.

jeremylt commented:
Yeah, I was thinking of leaving the 'raw' example to expose the details more easily for users.

jeremylt commented:
Merge button currently blocked by ReadTheDocs glitch

jeremylt commented:
@jedbrown I added the allowed ARM failure, this this branch is ready to merge.

CodeCov claims "A common practice in the modern developer's workflow is to test multiple build systems, split up tests in different containers, and group tests based on test focus (unittests vs integration tests). Codecov flawlessly supports all these setups. Therefore, no setup is required."
but that isn't actually working right, so there is a false decrease in coverage blocking the mergability of the branch now. I'm investigating but I don't know if/when I'll be able to fix.

jedbrown commented:
I've also had issues with that and the docs are useless for debugging. Is there a way to report bugs to codecov?

jeremylt commented:
They have a user forum. I'm planning on posting there asking for help when I get a chance.

jeremylt commented:
I rebased for the changes on main

jedbrown commented:
Is this branch going to be squashed?

jeremylt commented:
I was planning on doing one big squash.

jedbrown commented:
Go ahead as far as I'm concerned.

jeremylt commented:
Intel changed something and Travis is broken. Will investigate tomorrow

jeremylt commented:
I had a realization and was able to fix it.

jeremylt commented:
GCOV made big changes in the reports for GCC 8 (and then again for GCC 9). I'm trying to run Travis GCOV with GCC 8 like Noether has.

jeremylt commented:
OK, so it looks like isse with CodeCov is that GCC 8 is making wrong reports in our current configuration.

jedbrown commented:
Any reason not to use gcc-9 or gcc-10 or clang-10 for the C compiler?

jeremylt commented:
G Fortran 9\10 have the compiler bug.

jedbrown commented:
Ok, bummer that gcov can't use gfortran-8 with more recent gcc. So I guess we should just stick with everything from gcc-8 suite until the bug is fixed. I can do an admin merge when you're ready.

jeremylt commented:
I think I have a workaround for the Fortran issue, but I feel like I should play with this in a separate branch. I stripped those experimental commits. This is ready to squash+merge and I'll tinker on a different branch.
